src/platforms/linux/xdgstartatbootwatcher.cpp

```diff
@@ -0,0 +1,156 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "xdgstartatbootwatcher.h"
+
+#include <QDBusConnection>
+#include <QDBusMessage>
+#include <QDBusObjectPath>
+#include <QDBusPendingCallWatcher>
+#include <QDBusPendingReply>
+#include <QRandomGenerator>
+
+#if QT_VERSION >= QT_VERSION_CHECK(6, 5, 0)
+#  include <private/qgenericunixservices_p.h>
+#  include <private/qguiapplication_p.h>
+#  include <qpa/qplatformintegration.h>
+
+#  include "qmlengineholder.h"
+#else
+#  include <QGuiApplication>
+#endif
+
+#include "leakdetector.h"
+#include "logger.h"
+#include "settingsholder.h"
+
+namespace {
+Logger logger("XdgStartAtBootWatcher");
+}
+
+constexpr const char* XDG_PORTAL_SERVICE = "org.freedesktop.portal.Desktop";
+constexpr const char* XDG_PORTAL_PATH = "/org/freedesktop/portal/desktop";
+constexpr const char* XDG_PORTAL_BACKGROUND =
+    "org.freedesktop.portal.Background";
+constexpr const char* XDG_PORTAL_REQUEST = "org.freedesktop.portal.Request";
+
+XdgStartAtBootWatcher::XdgStartAtBootWatcher() : QObject() {
+  MZ_COUNT_CTOR(XdgStartAtBootWatcher);
+
+  logger.debug() << "StartAtBoot watcher";
+
+  connect(SettingsHolder::instance(), &SettingsHolder::startAtBootChanged, this,
+          &XdgStartAtBootWatcher::startAtBootChanged);
+
+  // Generate a unique token, and connect to the XDG Response portal.
+  quint64 randbits = QRandomGenerator::global()->generate64();
+  m_token = "mozillavpn_" + QString::number(randbits, 16);
+  m_replyPath = xdgReplyPath();
+  QDBusConnection::sessionBus().connect(XDG_PORTAL_SERVICE, m_replyPath,
+                                        XDG_PORTAL_REQUEST, "Response", this,
+                                        SLOT(xdgResponse(uint, QVariantMap)));
+
+  startAtBootChanged();
+}
+
+XdgStartAtBootWatcher::~XdgStartAtBootWatcher() {
+  MZ_COUNT_DTOR(XdgStartAtBootWatcher);
+}
+
+// See the org.freedesktop.portal.Request documentation, the request portal will
+// likely take the form "/org/freedesktop/portal/desktop/request/<SENDER>/TOKEN"
+// and we should connect to it before trying to make a request.
+QString XdgStartAtBootWatcher::xdgReplyPath() {
+  QDBusConnection bus = QDBusConnection::sessionBus();
+  QString sender = bus.baseService().mid(1).replace('.', '_');
+  return QString("/org/freedesktop/portal/desktop/request/%1/%2")
+      .arg(sender)
+      .arg(m_token);
+}
+
+void XdgStartAtBootWatcher::xdgResponse(uint response, QVariantMap results) {
+  logger.debug() << "StartAtBoot responded:" << response;
+  for (auto i = results.cbegin(); i != results.cend(); i++) {
+    logger.debug() << "StartAtBoot" << QString("%1:").arg(i.key())
+                   << i.value().toString();
+  }
+}
+
+// Try to find the window identifier for an XDG desktop portal request.
+// https://flatpak.github.io/xdg-desktop-portal/docs/window-identifiers.html
+QString XdgStartAtBootWatcher::parentWindow() {
```

Comment:
Consider making this `std::excpeted<const QString,SomeErrorEnum>` so the caller need's to check if this succeed.

Reply:
I think it's more in-line with the XDG specs to return an empty string here, as that is the expected value of the window handle in the event that the windowing system doesn't support it.

Reply:
I see thanks for explaining :)